tests/multi_bluetooth/ble_l2cap_coc.py

```diff
@@ -0,0 +1,170 @@
+    elif event == _IRQ_L2CAP_COC_TX_UNSTALLED:
+        pass
+
+    if event not in waiting_events:
+        waiting_events[event] = None
+
+
+def wait_for_event(event, timeout_ms):
+    t0 = time.ticks_ms()
+    while time.ticks_diff(time.ticks_ms(), t0) < timeout_ms:
+        if event in waiting_events:
+            return waiting_events.pop(event)
+        machine.idle()
+    raise ValueError("Timeout waiting for {}".format(event))
+
+
+def send_data(ble, chan_handle):
+    buf = bytearray(_PAYLOAD_LEN)
+    mv = memoryview(buf)
+    print("l2cap_coc_send", _NUM_PAYLOADS, _PAYLOAD_LEN)
+    for i in range(_NUM_PAYLOADS):
+        n = _PAYLOAD_LEN + i * _PAYLOAD_LEN_STEP
+        for j in range(n):
+            buf[j] = random.randint(0, 255)
+        if ble.l2cap_coc_send(chan_handle, mv[:n]):
+            wait_for_event(_IRQ_L2CAP_COC_TX_UNSTALLED, TIMEOUT_MS)
+
+
+def recv_data(ble, chan_handle):
+    buf = bytearray(_PAYLOAD_LEN)
+    recv_bytes = 0
+    recv_correct = 0
+    expected_bytes = (
+        _PAYLOAD_LEN * _NUM_PAYLOADS + _PAYLOAD_LEN_STEP * _NUM_PAYLOADS * (_NUM_PAYLOADS - 1) // 2
+    )
+    print("l2cap_coc_recvinto", expected_bytes)
+    while recv_bytes < expected_bytes:
+        wait_for_event(_IRQ_L2CAP_COC_RECV, TIMEOUT_MS)
+        while True:
+            n = ble.l2cap_coc_recvinto(chan_handle, buf)
+            if n == 0:
+                break
+            recv_bytes += n
+            for i in range(n):
+                if buf[i] == random.randint(0, 255):
+                    recv_correct += 1
+    return recv_bytes, recv_correct
+
+
+# Acting in peripheral role.
+def instance0():
+    multitest.globals(BDADDR=ble.config("mac"))
+    print("gap_advertise")
+    ble.gap_advertise(20_000, b"\x02\x01\x06\x04\xffMPY")
+    multitest.next()
+    try:
+        # Wait for central to connect to us.
+        conn_handle = wait_for_event(_IRQ_CENTRAL_CONNECT, TIMEOUT_MS)
+
+        print("l2cap_coc_listen")
+        ble.l2cap_coc_listen(_L2CAP_COC_PSM, _L2CAP_COC_MTU)
+
+        chan_handle = wait_for_event(_IRQ_L2CAP_COC_ACCEPT, TIMEOUT_MS)
+        chan_handle, our_mtu, peer_mtu = wait_for_event(_IRQ_L2CAP_COC_CONNECT, TIMEOUT_MS)
+
+        random.seed(_RANDOM_SEED)
+
+        recv_bytes, recv_correct = recv_data(ble, chan_handle)
+        send_data(ble, chan_handle)
+
+        wait_for_event(_IRQ_L2CAP_COC_DISCONNECT, TIMEOUT_MS)
+
+        print("received", recv_bytes, recv_correct)
+
+        # Wait for the central to disconnect.
+        wait_for_event(_IRQ_CENTRAL_DISCONNECT, TIMEOUT_MS)
+    finally:
+        ble.active(0)
+
+
+# Acting in central role.
+def instance1():
+    multitest.next()
+    try:
+        # Connect to peripheral and then disconnect.
+        print("gap_connect")
+        ble.gap_connect(*BDADDR)
+        conn_handle = wait_for_event(_IRQ_PERIPHERAL_CONNECT, TIMEOUT_MS)
+
+        print("l2cap_coc_connect")
+        ble.l2cap_coc_connect(conn_handle, _L2CAP_COC_PSM, _L2CAP_COC_MTU)
```

Comment:
is it possible to call the connect even if the other side is not yet listening?

Reply:
Yes. I've made this work by handling failed connections.